### PR TITLE
rgw: remove std:move() call that are unneeded

### PR DESCRIPTION
**Pull request by wjwithagen**
Clang warns and suggests in all these case allong this:
```
warning: moving a temporary object prevents copy elision [-Wpessimizing-move]
      return std::move(std::unique_ptr<RGWBucket>(new RGWRadosBucket(*this)));
             ^
/home/jenkins/workspace/ceph-master-compile/src/rgw/rgw_sal_rados.h:256:14: note: remove std::move call here
      return std::move(std::unique_ptr<RGWBucket>(new RGWRadosBucket(*this)));
             ^~~~~~~~~~                                                     ~
1 warning generated.
```

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/rgw/rgw_lc.cc [hunk 1]
```diff
@@ -29,10 +29,6 @@
 #include "rgw_sal.h"
 #include "rgw_sal_rados.h"
 
-// this seems safe to use, at least for now--arguably, we should
-// prefer header-only fmt, in general
-#undef FMT_HEADER_ONLY
-#define FMT_HEADER_ONLY 1
```
Comment (tchaikov) [inline src/rgw/rgw_lc.cc:35]: this change does not belong to this commit or probably this PR.

**Comment by mattbenjamin** [comment]
so, becaused clang says so, we're removing a bunch of potential optimizations

**Comment by mattbenjamin** [comment]
looking at the warnings themselves, we may be missing the opportunity to make a more appropriate optimization, in order to make clang's warning disappear

**Comment by mattbenjamin** [comment]
in other words, I think we should -not- do this

**Comment by wjwithagen** [comment]
@adamemerson is this also a set of moves that are there to prevent copying the data on purpose?

**Comment by wjwithagen** [comment]
@mattbenjamin 
Right I was reading this the other way around, because the compilers hand is forced into move() there is no way to select a possible alternative optimalisation. And that is what the compiler is complaining about.
It is now locked out of doing `copy elision`, which would mean that it is constructed at the place it would need to go. So it does not even have to move().
OTOH I saw @adamemerson suggest in #38308 hat some of the mov() are on purpose to commit to parts of the API contract.

**Comment by wjwithagen** [comment]
@mattbenjamin 
Even found some suggestions that Clang could be correct here:
[https://en.cppreference.com/w/cpp/language/copy_elision](url)

But formal language specification is rather tricky, and my English is nog good enough to get the the finer details here.

**Review by tchaikov — COMMENTED** [review]
i think all the `std::move()` removed by this PR should indeed by removed because as @wjwithagen quoted, copy elision ensures us the overhead of copy should be removed if the compiler complies to C++17. it's not like NVRO which is an optimization, which is optional from the compiler's perspective.

File: src/rgw/rgw_lc.cc (hunk 1 above)
Comment (wjwithagen) [reply src/rgw/rgw_lc.cc:35]: Nope, we had enough discussion about this one. ;-)
Looks like it is still in my patchset.

File: src/rgw/rgw_sal_rados.h [hunk 2]
```diff
@@ -253,7 +253,7 @@ class RGWRadosBucket : public RGWBucket {
 			   bool *is_truncated, RGWUsageIter& usage_iter,
 			   map<rgw_user_bucket, rgw_usage_log_entry>& usage) override;
     virtual std::unique_ptr<RGWBucket> clone() {
-      return std::move(std::unique_ptr<RGWBucket>(new RGWRadosBucket(*this)));
+      return std::unique_ptr<RGWBucket>(new RGWRadosBucket(*this));
```
Comment (tchaikov) [inline src/rgw/rgw_sal_rados.h:256]: nit, use `std::make_unique<>` instead? not strictly in the scope of this PR, though.

**Review by tchaikov — APPROVED** [review]

File: src/rgw/rgw_sal_rados.h (hunk 2 above)
Comment (wjwithagen) [reply src/rgw/rgw_sal_rados.h:256]: @tchaikov 
Let's do that in another PR, since a first try at this gives me lots of complaints about things like:
```
 note: unimplemented pure virtual method 'list' in 'RGWBucket'
    virtual int list(ListParams&, int, ListResults&, optional_yield y) = 0;
```
So that'll be more than just this simple replace.

**Review by adamemerson — APPROVED** [review]
This looks correct to me. In general moving the return value of a function blocks the opportunity for copy elision. (That is, the move keeps the compiler from simply constructing the value the function would return in the caller's stack, no move or copy at all.)

**Comment by wjwithagen** [comment]
> 
> 
> This looks correct to me. In general moving the return value of a function blocks the opportunity for copy elision. (That is, the move keeps the compiler from simply constructing the value the function would return in the caller's stack, no move or copy at all.)

Slightly off topic....
The only thing I'm wondering about  is how that works in something like this:

```
type_a f() { return A; }
....
type_a b = f();
```

Would the compiler actually be able to optimise it in such a way that `f()` directly writes into `b`, or that the assignment to `b` still introduces a `move`??

**Comment by adamemerson** [comment]
@wjwithagen Where's `A` coming from in that example?

**Comment by mattbenjamin** [comment]
@adamemerson this move clearly cannot be on the caller's stack.  The move is not actually eliding anything.  The larger issue is actually optimizing zipper handles.  The move actually indicated the author's intent, which this change is erasing.

**Comment by wjwithagen** [comment]
> @wjwithagen Where's `A` coming from in that example?

Good question, in my idea it was declared within `f()` or even created/constructed doing the return statement.
But you are right, it depends on the fact if A was created on the stack or on the heap. 
In the first case its lifespan ends with the exit from `f()` , and in the second case it will need GC once its use has gone.

**Comment by wjwithagen** [comment]
> 
> 
> @adamemerson this move (d object) clearly cannot be on the caller's stack. The move is not actually eliding anything. The larger issue is actually optimizing zipper handles. The move actually indicated the author's intent, which this change is erasing.

But it looks like the compiler does not agree? Note that this PR is to get ride of Clang complaining/warning....
I've never looked at generated ASM for C++ programs to see what code was generated, and decide which version is the better version. 

Alternative might be explain in a comment and surround with pragmas.....

**Comment by tchaikov** [comment]
> > This looks correct to me. In general moving the return value of a function blocks the opportunity for copy elision. (That is, the move keeps the compiler from simply constructing the value the function would return in the caller's stack, no move or copy at all.)
> 
> Slightly off topic....
> The only thing I'm wondering about is how that works in something like this:
> 
> ```
> type_a f() { return A; }
> ....
> type_a b = f();
> ```
> 
> Would the compiler actually be able to optimise it in such a way that `f()` directly writes into `b`, or that the assignment to `b` still introduces a `move`??

you probably need to use a more concrete example. for instance, see https://godbolt.org/z/dnPGvW . in which, the caller allocates 400 bytes on its stack and pass the address of the address of the 400-byte buffer to `makeA()`.

if we replace `return a` with `return std::move(a)`, an extra call of `std::move()` and an extra call of copy constructor are added.

anyway, Willem, if you really want to understand this. please read the C++ standard and the generated assembly. 

> > @wjwithagen Where's `A` coming from in that example?
> 
> Good question, in my idea it was declared within `f()` or even created/constructed doing the return statement.
> But you are right, it depends on the fact if A was created on the stack or on the heap.
> In the first case its lifespan ends with the exit from `f()` , and in the second case it will need GC once its use has gone.

i don't think this is the right way to explain this. neither do i think C++ has builtin GC.

**Review by dang — APPROVED** [review]
I'm fine with this.  Most of these are left over from my trying to get ceph::static_ptr to work, before I realized I needed to remove the copy case from it.

**Comment by mattbenjamin** [comment]
> 
> 
> I'm fine with this. Most of these are left over from my trying to get ceph::static_ptr to work, before I realized I needed to remove the copy case from it.

ok

**Comment by mattbenjamin** [comment]
@wjwithagen there are no more objections here, has there been a teuthology run?

**Comment by wjwithagen** [comment]
> 
> 
> @wjwithagen there are no more objections here, has there been a teuthology run?

I have not done any teuthology runs ever. Don't even have the rights to do so, I think.
Usually Kefu (@tchaikov) takes care of that. And I think this one sort got out of focus because of his holidays.

**Comment by tchaikov** [comment]
> > @wjwithagen there are no more objections here, has there been a teuthology run?
> 
> I have not done any teuthology runs ever. Don't even have the rights to do so, I think.
> Usually Kefu (@tchaikov) takes care of that. And I think this one sort got out of focus because of his holidays.

i am on vacation. but i am not qualified to investigate / analysis the test failures in rgw run. i had a handful failures when running it couple months ago.

**Comment by wjwithagen** [comment]
Overcome by time
